{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "d4062a5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#This performs a single peak MWHP conversion on an amino acid sequence\n",
    "def PCDTWConvert(x,PCProp1='Mass',PCProp2='HydroPho',normalize=False):\n",
    "    BaseArray=NoteValueConverter(PCProp1,PCProp2,Normalize=normalize)\n",
    "\n",
    "    #Create an empty list to hold the vectors produced from each run of the loop (JD).\n",
    "    tempnamesLST=[]\n",
    "    temptpLST=[]\n",
    "    tempspLST=[]\n",
    "  \n",
    "    currentseq = x\n",
    "    \n",
    "    lengther=len(currentseq)   \n",
    "    \n",
    "    #Convert all of the charaters to upper before converting it to a list (JD)\n",
    "    currentseq=currentseq.upper()\n",
    "    \n",
    "    #Convert the current sequence to a list (JD)\n",
    "    currentseq=list(currentseq)\n",
    "    \n",
    "    #Loop through all of the characters and if any are not allowed amino acids, replace them with a 'G'(AK)\n",
    "    for j in range(lengther):\n",
    "            strAllowedChars = [\"A\", \"C\", \"D\", \"E\", \"F\", \"G\", \"H\", \"I\", \"K\", \"L\",\"M\", \"N\", \"P\", \"Q\", \"R\", \"S\", \"T\", \"V\", \"W\", \"Y\"]\n",
    "    \n",
    "            if currentseq[j] not in strAllowedChars: \n",
    "                currentseq[j] = currentseq[j].replace(currentseq[j],'G')\n",
    "    \n",
    "    #Convert the list back to a string (JD)\n",
    "    currentseq=''.join(currentseq)\n",
    "   \n",
    "    ##This block of code creates a vector containing only the notevalues representing the amino acid sequence (JD).    \n",
    "    NoteVec=[]\n",
    "    ender=currentseq[-1]\n",
    "    currentseq=\"G\"+currentseq+ender\n",
    "    lengther=len(currentseq)\n",
    "    \n",
    "    for k in range(lengther):\n",
    "        if k==lengther:\n",
    "            NoteVec.append(BaseArray[currentseq[k-1]].NoteValue)\n",
    "        else:\n",
    "            NoteVec.append(BaseArray[currentseq[k]].NoteValue)\n",
    "  \n",
    "    ##This block of code converts a vector of NoteValues to a single-peak vector (JD)   \n",
    "    lengther2=(len(NoteVec))-1\n",
    "    SinglePeakOutVector=[]\n",
    "    \n",
    "    for l in range(lengther2):\n",
    "        if l==0:\n",
    "            Outvalue=((NoteVec[l]*2)+((NoteVec[l]+NoteVec[l+1])/2))/2\n",
    "            SinglePeakOutVector.append(round(Outvalue,4))\n",
    "        else:\n",
    "            Outvalue=((NoteVec[l])+((NoteVec[l]+NoteVec[l+1])/2)+((NoteVec[l]+NoteVec[l-1])/2))/2  \n",
    "            SinglePeakOutVector.append(round(Outvalue,4))\n",
    "\n",
    "    return SinglePeakOutVector[1:], NoteVec\n",
    "\n",
    "#Conversion function that takes any two PC properties and converts them to a notevalue list\n",
    "def NoteValueConverter(PropA,PropB,Normalize=True):\n",
    "    \n",
    "    import pandas as pd\n",
    "    \n",
    "    #The values in this dataframe were taken from DOI:10.1371/journal.pone.0147 and vary slightly from the values used in\n",
    "    #The MWHP algorithm used in Dixson and Azad 2021.  Zero values were treated as 0.001 here except for hydrophobicity\n",
    "    #where they were set to 1 as in the original algorithm.  The values for hydrophobicity are mostly from Monera 1995 except for \n",
    "    #proline that was not included in that publication.  The value for proline was set at -46 in accordance with the chart at\n",
    "    #https://www.sigmaaldrich.com/US/en/technical-documents/technical-article/protein-biology/protein-structural-analysis/amino-acid-reference-chart?srsltid=AfmBOop298sG6HwDyQ7tvsDFLsBOA2B9f8MQMalLOrTFF3h3_fC7iqZC\n",
    "    #which includes the values from Monera et al 1995.\n",
    "\n",
    "    \n",
    "    PCPropertiesRaw=pd.DataFrame.from_dict({'HydroPho':[1,41,97,74,100,97,-23,-10,-31,-5,-46,76,99,49,63,8,-14,-28,-55,13],\n",
    "                                    'HydroPhIl':[0.001,-0.5,-1.8,-1.3,-2.5,-3.4,3,0.2,3,0.3,0.001,-1.5,-1.8,-1,-2.3,-0.5,3,2,3,-0.4],\n",
    "                                    'Hbond':[2,2,2,2,2,3,2,4,4,4,2,2,2,2,3,4,4,4,4,4],\n",
    "                                    'SideVol':[0.001,27.5,93.5,94.1,115.5,145.5,100,80.7,62,29.3,41.9,71.5,93.5,44.6,117.3,79,105,58.7,40,51.3],\n",
    "                                    'Polarity':[9,8.1,4.9,5.7,5.2,5.4,11.3,10.5,12.3,9.2,8,5.9,5.2,5.5,6.2,10.4,10.5,11.6,13,8.6],\n",
    "                                    'Polarizability':[0.001,0.046,0.186,0.221,0.29,0.409,0.219,0.18,0.151,0.062,0.131,0.14,0.186,0.128,0.298,0.23,0.18,0.134,0.105,0.108],\n",
    "                                    'SASA':[0.881,1.181,1.931,2.034,2.228,2.663,2.258,1.932,1.862,1.298,1.468,1.645,1.81,1.461,2.368,2.025,1.932,1.655,1.587,1.525],\n",
    "                                    'NCI':[0.179052,0.007187,0.051672,0.002683,0.037552,0.037977,0.017708,0.049211,0.006802,0.004627,0.239531,0.057004,0.021631,-0.03661,0.023599,-0.01069,0.049211,0.005392,-0.02382,0.003352],\n",
    "                                    'Mass':[57.05,71.08,113.16,131.2,147.18,186.22,128.18,128.13,129.12,87.08,97.12,99.13,113.16,103.15,163.18,137.14,156.19,114.11,115.09,101.11]},\n",
    "                                    orient='index', columns=['G','A','L','M','F','W','K','Q','E','S','P','V','I','C','Y','H','R','N','D','T'])\n",
    "    \n",
    "    PCPropertiesNorm=pd.DataFrame()\n",
    "    PCProperties=pd.DataFrame()\n",
    "\n",
    "    def absolute_maximum_scale(series):\n",
    "        return series / series.abs().max()\n",
    "\n",
    "    PCPropertiesNorm=PCPropertiesRaw.transpose()\n",
    "\n",
    "    for col in PCPropertiesNorm.columns:\n",
    "        PCPropertiesNorm[col] = absolute_maximum_scale(PCPropertiesNorm[col])\n",
    "    PCPropertiesNorm=PCPropertiesNorm.transpose() \n",
    "    \n",
    "    if Normalize is True:\n",
    "        PCProperties=PCPropertiesNorm\n",
    "    else:\n",
    "        PCProperties=PCPropertiesRaw\n",
    "     \n",
    "    PreNotesA=PCProperties.loc[PropA].tolist()\n",
    "    PreNotesB=PCProperties.loc[PropB].tolist()\n",
    "    \n",
    "    MaxNoteA=max(PreNotesA)\n",
    "    \n",
    "    outNoteslst=[]\n",
    "    \n",
    "    for i in range(20):\n",
    "        R=PCProperties.loc[PropA].tolist()[i]\n",
    "        G=max(PreNotesA)\n",
    "        H=PCProperties.loc[PropB].tolist()[i]\n",
    "        V=round(((R/G)*H),4)\n",
    "        outNoteslst.append(V)\n",
    "        \n",
    "    BaseArray=pd.DataFrame.from_dict({\n",
    "        'NoteValue': outNoteslst},\n",
    "    orient='index', columns=['G','A','L','M','F','W','K','Q','E','S','P','V','I','C','Y','H','R','N','D','T'])\n",
    "   \n",
    "    return BaseArray\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4b1fb1ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "#This performs a single peak MWHP alignment\n",
    "def PCDTWAlign(inputseq1str,inputseq2str,PCProp1='Mass',PCProp2='HydroPho',Penalty=0,Window=3):\n",
    "    from dtaidistance import dtw\n",
    "    import numpy as np\n",
    "\n",
    "    vector1=PCDTWConvert(inputseq1str,PCProp1,PCProp2)\n",
    "    vector2=PCDTWConvert(inputseq2str,PCProp1,PCProp2)\n",
    "\n",
    "    path=dtw.warping_path(vector1, vector2,penalty=Penalty,window=Window)\n",
    "    \n",
    "    #Split the path into two lists that represent index values of the original sequence.\n",
    "    vector1indexlst, vector2indexlst = zip(*path)\n",
    "    \n",
    "    alignedseq1=\"\"\n",
    "    alignedseq2=\"\"\n",
    "    \n",
    "    h=0\n",
    "    while h <(len(vector1indexlst)):\n",
    "        looker1=vector1indexlst[h]\n",
    "        temp1counter=vector1indexlst.count(looker1)\n",
    "        if temp1counter==1:\n",
    "            alignedseq1=alignedseq1+(list(inputseq1str))[vector1indexlst[h]]\n",
    "            h+=temp1counter\n",
    "        elif temp1counter>1:\n",
    "            alignedseq1=alignedseq1+(list(inputseq1str))[vector1indexlst[h]]\n",
    "            for l in range(temp1counter-1):\n",
    "                alignedseq1=alignedseq1+\"-\"\n",
    "            h+=temp1counter\n",
    "    \n",
    "    \n",
    "    j=0\n",
    "    while j <(len(vector2indexlst)):\n",
    "        looker2=vector2indexlst[j]\n",
    "        temp2counter=vector2indexlst.count(looker2)\n",
    "        if temp2counter==1:\n",
    "            alignedseq2=alignedseq2+(list(inputseq2str))[vector2indexlst[j]]\n",
    "            j+=temp2counter\n",
    "        elif temp2counter>1:\n",
    "            alignedseq2=alignedseq2+(list(inputseq2str))[vector2indexlst[j]]\n",
    "            for l in range(temp2counter-1):\n",
    "                alignedseq2=alignedseq2+\"-\"\n",
    "            j+=temp2counter\n",
    "            \n",
    "            \n",
    "###########################GET THE ALIGNED VECTORS########################################\n",
    "    alignedvec1=[]\n",
    "    alignedvec2=[]\n",
    "    \n",
    "    h=0\n",
    "    while h <(len(vector1indexlst)):\n",
    "        looker1=vector1indexlst[h]\n",
    "        temp1counter=vector1indexlst.count(looker1)\n",
    "        if temp1counter==1:\n",
    "            alignedvec1.append(vector1[vector1indexlst[h]])\n",
    "            h+=temp1counter\n",
    "        elif temp1counter>1:\n",
    "            alignedvec1.append(vector1[vector1indexlst[h]])\n",
    "            for l in range(temp1counter-1):\n",
    "                alignedvec1.append(vector1[vector1indexlst[h]])\n",
    "            h+=temp1counter\n",
    "    \n",
    "    \n",
    "    j=0\n",
    "    while j <(len(vector2indexlst)):\n",
    "        looker2=vector2indexlst[j]\n",
    "        temp2counter=vector2indexlst.count(looker2)\n",
    "        if temp2counter==1:\n",
    "            alignedvec2.append(vector2[vector2indexlst[j]])\n",
    "            j+=temp2counter\n",
    "        elif temp2counter>1:\n",
    "            alignedvec2.append(vector2[vector2indexlst[j]])\n",
    "            for l in range(temp2counter-1):\n",
    "                alignedvec2.append(vector2[vector2indexlst[j]])\n",
    "            j+=temp2counter\n",
    "    outp1arr=np.array(alignedvec1)\n",
    "    outp2arr=np.array(alignedvec2)\n",
    "    ConsensusVector=(outp1arr+outp2arr)/2\n",
    "            \n",
    "###########################GET THE ALIGNED VECTORS########################################\n",
    "    \n",
    "    \n",
    "    GAPS=(alignedseq1+alignedseq2).count('-')\n",
    "    Seq1Alignstr=alignedseq1\n",
    "    Seq2Alignstr=alignedseq2\n",
    "    \n",
    "    counter=0\n",
    "    for i in range(len(Seq1Alignstr)):\n",
    "        if Seq1Alignstr[i]==Seq2Alignstr[i]:\n",
    "            counter=counter+1\n",
    "    identity=(int((counter/len(Seq1Alignstr))*100))\n",
    "    \n",
    "    midalignstr=''\n",
    "    for i in range(len(Seq1Alignstr)):\n",
    "        if Seq1Alignstr[i]==Seq2Alignstr[i]:\n",
    "            midalignstr=midalignstr+'|'\n",
    "        else:\n",
    "            midalignstr=midalignstr+' '\n",
    "    FullAlignment= Seq1Alignstr+'\\n'+midalignstr+'\\n'+Seq2Alignstr\n",
    "    \n",
    "    #Full list of variables that could be returned:Seq1Alignstr, Seq2Alignstr, GAPS, path, identity, alignedvec1, alignedvec2, ConsensusVector\n",
    "    PCDTWAlnResult={\n",
    "        'Seq1AlignedString':Seq1Alignstr,\n",
    "        'Seq2AlignedString':Seq2Alignstr,\n",
    "        'FullAlignment':FullAlignment,\n",
    "        'Identity':identity,\n",
    "        'ConsensusVector':ConsensusVector       \n",
    "                   }\n",
    "\n",
    "    return PCDTWAlnResult"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2fe0195d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#This returns the dtw distance per residue.  Default is single peak vector (i.e. one peak per residue)\n",
    "def PCDTWDist(Seq1,Seq2,PCProp1='Mass',PCProp2='HydroPho',PeaksPerRes=1,Penalty=0,Window=3):\n",
    "    from dtaidistance import dtw\n",
    "    Vec1=PCDTWConvert(Seq1,PCProp1,PCProp2)\n",
    "    Vec2=PCDTWConvert(Seq2,PCProp1,PCProp2)\n",
    "    \n",
    "    DTWDistance = dtw.distance(Vec1, Vec2, penalty=Penalty, window=Window)\n",
    "    DTWDistance=DTWDistance/(len(Vec1)*PeaksPerRes)\n",
    "    DTWDistance=round(DTWDistance,3)\n",
    "\n",
    "    return DTWDistance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "ed6eb4d5-4abd-45a3-a3ce-2d65aaaa3234",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">ProtX_SimHomolog_1\n",
      "GAIM\n",
      ">ProtX_SimHomolog_2\n",
      "GAIY\n",
      ">ProtX_SimHomolog_3\n",
      "GTIY\n",
      ">ProtX_SimHomolog_4\n",
      "STIY\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "#Function to evolve a sequence with the objective of minimizing the physicochemical change (2 properties) while maximizing the sequence level change.\n",
    "#The sequences output from this function will vary slightly from those Dixson et. al. 2025 because the hydrophobicity values have been updated\n",
    "#This change is discussed further in the NoteValueConverter function comments above.\n",
    "\n",
    "def PCEvolve(Seq='GALM', PCProp1='Mass', PCProp2='HydroPho', BaseName='ProtX'):\n",
    "    import pandas as pd\n",
    "    NoteVector = (PCDTWConvert(x=Seq, PCProp1=PCProp1, PCProp2=PCProp2, normalize=False))[1]\n",
    "    NoteVector=NoteVector[:-1]\n",
    "    NoteVector=NoteVector[1:]\n",
    "\n",
    "    #Create an empty dataframe to hold the top three residues that match the input sequence\n",
    "    MatchedRes=pd.DataFrame()      \n",
    "    \n",
    "    # #Creat lists to hold the residues and the notevalue for each residue\n",
    "    NoteResLst=(NoteValueConverter(PropA=PCProp1,PropB=PCProp2,Normalize=False)).columns.tolist()  \n",
    "    NoteValLst=(NoteValueConverter(PropA=PCProp1,PropB=PCProp2,Normalize=False)).iloc[0].tolist()\n",
    "\n",
    "    #Put the Notevalues into a dict\n",
    "    NoteDict = dict(zip(NoteResLst, NoteValLst))\n",
    "\n",
    "    #Convert the input sequence to a vector of note values\n",
    "    NoteValues = [NoteDict[char] for char in Seq if char in NoteDict]\n",
    "\n",
    "    #Find the closest fully divergent sequence to the input NoteValues\n",
    "    FullDivSeq = []\n",
    "    for value in NoteValues:\n",
    "        closest_key = None\n",
    "        min_difference = float('inf')\n",
    "        \n",
    "        for key, dict_value in NoteDict.items():\n",
    "            difference = abs(dict_value - value)\n",
    "            \n",
    "            if difference < min_difference and dict_value != value:\n",
    "                min_difference = difference\n",
    "                closest_key = key\n",
    "        \n",
    "        if closest_key is not None:\n",
    "            FullDivSeq.append(closest_key)\n",
    "\n",
    "    #Iterate through the NoteValues 20 times with a step of 0.05 difference so that a residue is only replaced if it is the step percentage value\n",
    "    #or less different than the original sequence\n",
    "    fasta_output = \"\"\n",
    "    threshold = 0  # Starts at 0% for identical sequence\n",
    "    previous_seq = Seq  # Initialize with original sequence to exclude it\n",
    "    seq_number = 1  # Output sequence numbering\n",
    "    \n",
    "    for i in range(20):\n",
    "        new_seq = \"\"\n",
    "    \n",
    "        for j, char in enumerate(Seq):\n",
    "            if char in NoteDict:\n",
    "                original_value = NoteDict[char]\n",
    "                alternative_value = NoteDict.get(FullDivSeq[j], original_value)  # Alternative or original value\n",
    "                percent_difference = abs(alternative_value - original_value) / original_value * 100\n",
    "    \n",
    "                # Final iteration: Apply 95% threshold and include all values exceeding it\n",
    "                if i == 19 and percent_difference > 95:\n",
    "                    new_seq += FullDivSeq[j]\n",
    "                elif percent_difference <= threshold and alternative_value != original_value:\n",
    "                    new_seq += FullDivSeq[j]\n",
    "                else:\n",
    "                    new_seq += char\n",
    "            else:\n",
    "                new_seq += char  # Preserve unchanged if not in dictionary\n",
    "    \n",
    "        # Append sequence only if it's different from the previous one AND not the original Seq\n",
    "        if new_seq != previous_seq:\n",
    "            fasta_output += f\">{BaseName}_{'SimHomolog_'}{seq_number}\\n{new_seq}\\n\"\n",
    "            previous_seq = new_seq  # Update sequence tracker\n",
    "            seq_number += 1  # Increment naming only for unique sequences\n",
    "    \n",
    "        # Increase threshold by 5% per iteration, stopping at 95% for the final loop\n",
    "        threshold = min(threshold + 5, 95)\n",
    "       \n",
    "    return fasta_output\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
